README.md
# BLZ&mdash;Bankleitzahlen in Ruby

## Description

BLZ or Bankleitzahlen is a bank identifier code system used by German
and Austrian banks (see [Wikipedia](https://en.wikipedia.org/wiki/Bankleitzahl)).

This gem provides for searching and returning the information
represented by the BLZ, such as the name of the bank etc.

## Installation

    $ gem install blz

## Command line tools

I provided a command line tool to search for BLZ:

```sh
$ blz exact 70150000
70150000, Stadtsparkasse München, 80791 München, SSKMDEMMXXX
```

Or search for BIC matches:

```sh
$ blz bic SSKMDE
70150000, Stadtsparkasse München, 80791 München, SSKMDEMMXXX
```

Or use substring matches:

```sh
$ blz match 4945
49450120, Sparkasse Herford, 32045 Herford, WLAHDE44XXX
49450120, Sparkasse Herford, 32285 Rödinghausen, Westf
49450120, Sparkasse Herford, 32132 Spenge
49450120, Sparkasse Herford, 32269 Kirchlengern
49450120, Sparkasse Herford, 32122 Enger, Westf
49450120, Sparkasse Herford, 32591 Vlotho
49450120, Sparkasse Herford, 32120 Hiddenhausen
49450120, Sparkasse Herford, 32221 Bünde
49450120, Sparkasse Herford, 32556 Löhne
49451210, Sparkasse Bad Salzuflen -alt-, 32102 Bad Salzuflen, WELADED1BSU
```

Or find by city (rather uncommon, but useful):

```sh
$ blz city Münstereifel
37069627, Raiffeisenbank Rheinbach Voreifel, 53902 Bad Münstereifel
37070024, Deutsche Bank Privat und Geschäftskunden, 53902 Bad Münstereifel, DEUTDEDB379
37070060, Deutsche Bank, 53902 Bad Münstereifel, DEUTDEDK379
38250110, Kreissparkasse Euskirchen, 53896 Bad Münstereifel
38260082, Volksbank Euskirchen, 53895 Bad Münstereifel
```

## Downloads

You can download the current list of BLZ (free of charge)
[from the German Bundesbank](http://www.bundesbank.de/Redaktion/DE/Standardartikel/Kerngeschaeftsfelder/Unbarer_Zahlungsverkehr/bankleitzahlen_download.html),
however, you need to complete their [registration process](https://extranet.bundesbank.de/bsvpub/).

The Bundesbank releases a new list every three months. The last part of the
version number of this gem mirrors the last date the provided data is valid.

Now go and build your own BLZ gem ;-)

### Note on converting

If you have an Extranet account, you may use the `scripts/fetch.rb` helper,
which will perform login, download, conversion (and logout). You need to
provide it with your credentials, like so:

```sh
$ EXTRANET_USERNAME="..." EXTRANET_PASSWORD="..." scripts/fetch.rb
```

It will create a new `data/*.tsv.gz` file or do nothing if the download
provided already exists.

Note, that you'll need to clone this repository and execute `bundle` in
the repository directory to fetch development dependencies.

## Contributors

* [dmke](https://github.com/dmke)
* [max-power](https://github.com/max-power)

## License

Released under the MIT License. See the LICENSE file for further
details.



CHANGELOG.md
# 0.2.0.20190603, releases 2019-05-03

* BLZ data file updated
* the previous cycle (20190304) did not contain updates

# 0.2.0.20181203, released 2018-11-01

* BLZ data file updated
* continued to fix automation scripts (again, does not affect library code)

# 0.2.0.20180903, released 2018-10-16

* BLZ data file updated
* small bugfix in automation scripts (does not affect BLZ library)

# 0.2.0.20180604, released 2018-05-04

* BLZ data file for cycle 2018-06-04 to 2018-08-02 added

# 0.2.0.20180305, released 2018-03-16

* BLZ data file for cycle 2018-03-05 to 2018-06-03 added

# 0.2.0.20171204, released 2017-05-08

* BLZ data file for cycle 2017-12-04 to 2018-03-04 added
* previous data set (2017-09-03 to 2017-12-04) was missed

# 0.2.0.20170605, released 2017-05-08

* BLZ data file for cycle 2017-06-04 to 2017-09-03 added
* BLZ data files for 2015 removed
* revamped tests, so that manual editing becomes unnecessary

# 0.1.4.20170306, released 2017-02-06

* BLZ data file for cycle 2017-03-06 to 2017-06-04 added
* data for 2016-09-04 until 2017-03-06 is unavailable, sorry

# 0.1.4.20160904, released 2016-05-15

* BLZ data file for cycle 2016-03-06 to 2016-09-04 added

# 0.1.4.20160306, released 2015-12-04

* BLZ data file for cycle 2015-12-07 to 2016-03-06 added
* removed data files for 2012-2014
* refactored data file finder logic

# 0.1.4.20151206, released 2015-08-19

* BLZ data file for cycle 2015-09-07 to 2015-12-06 added

# 0.1.4.20150607, released 2015-03-09

* BLZ data file for cycle 2015-06-09 to 2015-09-07 added

# 0.1.4.20141207, released 2014-09-08

* pinned data to last update cycle (there were no changes)
  This only gets rid of the "[BLZ] The data provided may not be accurate."
  warning.

# 0.1.4.20140907.2, released 2014-07-25

* finders return an empty array when given `nil` or a blank string

# 0.1.4.20140907.1, released 2014-07-11

* apply update BIC of Sberbank (SEZBDEF1XXX → SEZDDEF1XXX)
  after Bundesbank ExtraNet notifcation from 2014-07-10

# 0.1.4.20140907, released 2014-06-26

* BLZ data file for cycle 2014-06-09 to 2014-09-07 added
* generalized finder for `BLZ::DATA_FILE`

# 0.1.4.20140608, released 2014-02-13

* BLZ data file for cycle 2014-03-03 to 2014-06-08 added
* compressed data files for smaller gem size
* added `BLZ::Bank.find_by_bic` method
* removed XLSX source files

# 0.1.3, released 2013-02-12

* BLZ data file for cycle 2013-12-09 to 2014-03-02 added

# 0.1.2, released 2012-11-18

* BLZ data file not correctly resolved when
  using blz script

# 0.1.1, released 2012-11-18

* Fixes a bug with `nil` values (kudos to @max-power)
* Refactoring code, and it's just better to
  read now (again, thanks to @max-power)

# 0.1.0, released 2012-09-27

* Initial version released on GitHub


test/test_bank.rb
# encoding: utf-8

require 'blz'
require 'test/unit'
require 'stringio'

class TestBank < Test::Unit::TestCase
  DATA_FILES = Dir[File.join(File.dirname(__FILE__), '../data/*.tsv.gz')].sort

  DATA_FILES.each do |filename|
    define_method "test_find_file_for_#{File.basename(filename)[0..-8]}" do
      assert_stderr_no_match "The data provided may not be accurate" do
        date = BLZ.convert_file_to_date(filename)
        file = BLZ.find_data_file(date)
        assert_equal b(filename), b(file)
      end
    end
  end

  def test_find_file_for_past_date
    assert_stderr_match "The data provided may not be accurate" do
      file  = DATA_FILES.first
      date  = BLZ.convert_file_to_date(file) - 1
      found = BLZ.find_data_file(date)
      assert_equal b(file), b(found)
    end
  end

  def test_find_file_for_future_date
    assert_stderr_match "The data provided may not be accurate" do
      file  = DATA_FILES.last
      date  = BLZ.convert_file_to_date(file) + 91
      found = BLZ.find_data_file(date)
      assert_equal b(file), b(found)
    end
  end

  def test_current_data_file
    re = /\A\d{4}_\d\d_\d\d\.tsv\.gz\z/
    assert_match re, File.basename(BLZ::DATA_FILE)
  end

  def test_all_banks
    assert !BLZ::Bank.all.nil?
    assert sskm = BLZ::Bank.all.find { |b| b.blz == "70150000" }
    assert_equal "Stadtsparkasse München", sskm.name
  end

  def test_search_by_exact_match_munich
    assert results = BLZ::Bank.find_by_blz("70150000", :exact => true)
    assert_equal 1,                         results.size
    assert_equal "70150000",                results[0].blz
    assert_equal "Stadtsparkasse München",  results[0].name
    assert_equal "Stadtsparkasse München",  results[0].short_name
    assert_equal "80791",                   results[0].zip
    assert_equal "München",                 results[0].city
    assert_equal "SSKMDEMMXXX",             results[0].bic
  end

  def test_search_by_exact_match_bremen
    assert results = BLZ::Bank.find_by_blz("29050101", :exact => true)
    assert_equal 2,                   results.size
    assert_equal "29050101",          results[0].blz
    assert_equal "Sparkasse Bremen",  results[0].name
    assert_equal "Spk Bremen",        results[0].short_name
    assert_equal "28078",             results[0].zip
    assert_equal "Bremen",            results[0].city
    assert_equal "SBREDE22XXX",       results[0].bic
  end

  def test_search_by_prefix
    assert results = BLZ::Bank.find_by_blz("7106")
    assert results.size > 1
    #assert_equal "Stadtsparkasse München", results[0].bezeichnung
  end

  def test_search_by_city
    assert results = BLZ::Bank.find_by_city("München")
    assert results.size > 1
    assert results.find { |b| b.blz == "70150000" }
  end

  def test_search_by_bic
    assert results = BLZ::Bank.find_by_bic("SSKMDEMMXXX")
    assert results.size == 1
    assert results.find { |b| b.blz == "70150000" }
  end

  def test_search_by_empty_bic
    assert results = BLZ::Bank.find_by_bic('')
    assert results.size == 0
  end

  def test_search_by_nil_bic
    assert results = BLZ::Bank.find_by_bic(nil)
    assert results.size == 0
  end

  def test_to_s
    assert results = BLZ::Bank.find_by_blz("70150000", :exact_only => true)
    assert_equal 1, results.size
    assert_equal "70150000, Stadtsparkasse München, 80791 München, SSKMDEMMXXX", results[0].to_s
  end

  # helper

  def assert_stderr_match(msg, &block)
    serr = redirect_stderr(&block)
    assert_match msg, serr
  end

  def assert_stderr_no_match(msg, &block)
    msg = Regexp.new(Regexp.escape(msg)) unless msg.is_a?(Regexp)
    serr = redirect_stderr(&block)
    assert_no_match msg, serr
  end

  def b(file)
    File.basename file
  end

  def redirect_stderr
    fake_stderr = StringIO.new
    original_stderr, $stderr = $stderr, fake_stderr
    yield
    fake_stderr.string
  ensure
    $stderr = original_stderr
  end
end


lib/blz.rb
# Copyright (C) 2012 Oliver Eilhard
#
# This library is freely distributable under the terms of
# an MIT-style license.
# See COPYING or http://www.opensource.org/licenses/mit-license.php.
#
# Provides information about "Bankleitzahlen" (BLZ),
# a bank identifier code system used by German and Austrian banks.

require "date"

module BLZ
  def self.convert_file_to_date(f)
    match = f.match(/(?<y>\d{4})_(?<m>\d\d)_(?<d>\d\d)\.tsv\.gz$/)
    Date.new match[:y].to_i, match[:m].to_i, match[:d].to_i
  end

  def self.find_data_file(now=Date.today)
    glob = Dir[ File.join(File.dirname(__FILE__), '../data/*.tsv.gz') ].sort
    file = glob.find {|c| now <= convert_file_to_date(c) } || glob.last

    # sanity check
    if now < Date.new(2016, 3, 6) || now > (convert_file_to_date(file) + 90)
      warn [now]
      warn '[BLZ] The data provided may not be accurate.'
    end

    file
  end

  DATA_FILE = find_data_file
end

require 'blz/bank'


lib/blz/bank.rb
# encoding: utf-8
require 'csv'
require 'zlib'

module BLZ
  class Bank
    #  Represents a bank with a BLZ, a BIC, a name etc.
    class << self
      # Returns an array of all Banks specified by +code+.
      # The following options apply:
      #    exact:: Only return exact matches (false by default)
      #
      def find_by_blz(code, options = {})
        return [] if blank?(code)

        exact = options.fetch(:exact, false)
        all.select do |bank|
          bank.blz == code || (!exact && bank.blz.start_with?(code))
        end
      end

      # Returns an array of all Banks with a substring of +city+.
      def find_by_city(substring)
        return [] if blank?(substring)

        all.select do |bank|
          bank.city.index(substring)
        end
      end

      # Returns an array of all Banks specified by +bic+ (Business Identifier Code).
      # The following options apply:
      #    exact:: Only return exact matches (false by default)
      #
      def find_by_bic(bic, options = {})
        return [] if blank?(bic)

        exact = options.fetch(:exact, false)
        all.select do |bank|
          bank.bic == bic || (!exact && (bank.bic || '').start_with?(bic))
        end
      end

      # Returns an array of all banks.
      def all
        @banks ||= read_banks
      end

      private

      def read_banks
        banks = []
        Zlib::GzipReader.open(BLZ::DATA_FILE) do |gz|
          CSV.new(gz, col_sep: "\t").each do |r|
            banks << new(r[0], r[2], r[3], r[4], r[5], r[7])
          end
        end
        banks
      end

      # Checks whether an object is blank (empty Array/Hash/String).
      #
      # Does not rely on ActiveSupport, but prefers that implementation.
      def blank?(obj)
        return true       if obj.nil?
        return obj.blank? if obj.respond_to?(:blank?)
        obj = obj.strip   if String === obj
        return obj.empty? if obj.respond_to?(:empty?)

        false
      end
    end

    attr_reader :blz, :name, :zip, :city, :short_name, :bic

    # Initializes a single Bank record.
    def initialize(blz, name, zip, city, short_name, bic)
      @blz  = blz
      @name = name
      @zip  = zip
      @city = city
      @bic  = bic
      @short_name = short_name
    end

    # Returns a nice representation of the bank.
    def to_s
      [blz, name, "#{zip} #{city}", bic].compact.reject(&:empty?).join(', ')
    end

  end
end


scripts/convert.rb
#!/usr/bin/env ruby
#
# Helper to convert an Extranet BLZ file download in a more sane format.
#

if ARGV[0].nil?
  $stderr.puts "Usage: #{$0} FILE|-"
  exit 1
end

# length (and name as comment) of each column
METRIC = [
  8,  # BLZ
  1,  # Merkmal
  58, # Bezeichnung
  5,  # PLZ
  35, # Ort
  27, # Kurzbezeichnung
  5,  # PAN
  11, # BIC
  2,  # Prüfzifferberechnungsmethode
  6,  # Datensatznummer
  1,  # Änderungskennzeichen
  1,  # BLZ-Löschung
  8,  # Nachfolge-BLZ
]

METRIC_LEN = METRIC.inject(:+)

# read from stdin if arg is "-", and setup encoding transformation
SOURCE = (ARGV[0] == "-" ? $stdin : File.open(ARGV[0], "rb"))
  .set_encoding(Encoding::ISO_8859_15, Encoding::UTF_8)

while line = SOURCE.gets
  line.chomp!
  if line.length != METRIC_LEN
    $stderr.puts "expected line length #{METRIC_LEN}, got #{line.length} in '#{line}'"
    next
  end

  i = 0
  puts METRIC.inject([]) {|s, m|
    # extract columns, and strip leading/trailing whitespace
    s << line[i ... (i+=m)].strip
  }.join("\t")
end


scripts/fetch.rb
#!/usr/bin/env ruby

require "mechanize"
require "logger"
require "zlib"
require "zip"

if (f = File.expand_path("~/.extranet")) && File.exist?(f)
  username, password = File.read(f).strip.split(":", 2)
else
  username = ENV.fetch("EXTRANET_USERNAME")
  password = ENV.fetch("EXTRANET_PASSWORD")
end

# from ./convert.rb
METRIC = [8, 1, 58, 5, 35, 27, 5, 11, 2, 6, 1, 1, 8]
METRIC_LEN = METRIC.inject(:+)

NAVIGATION = {
  login:  "https://extranet.bundesbank.de/FT/",
  table:  "https://extranet.bundesbank.de/FT/dirHTML.do?dirKind=DOWN&ftclient=browser&tableView=kompakt",
  logout: "https://extranet.bundesbank.de/FT/logout.do",
}

# don't double download already existing files
last = Dir[ File.expand_path('../data/*.tsv.gz', __dir__) ].sort.last
last_match = last.match(/(?<y>\d{4})_(?<m>\d\d)_(?<d>\d\d)\.tsv\.gz$/)

logger = Logger.new($stderr).tap{|l| l.level = Logger::INFO }
agent = Mechanize.new do |m|
  # m.log = logger
  m.user_agent_alias = "Windows IE 11"
  m.pluggable_parser["application/octet-stream"] = Mechanize::Download
end

begin
  logger.info("login")
  page = agent.get NAVIGATION[:login]
  form = page.form_with action: "/pkmslogin.form"
  form["username"] = username
  form["password"] = password
  agent.submit(form)

  logger.info("get list of downloads")
  sleep 1
  page = agent.get NAVIGATION[:table]

  target_link = page.links
    .select  {|link| !!(link.text.strip =~ /BLZ_\d{8}.zip/) }
    .sort_by {|link| link.text.strip }
    .last

  if !target_link
    logger.info("no new download found")
    exit 0
  end

  name = target_link.text.strip
  if name == format("BLZ_%4d%02d%02d.zip", last_match[:y].to_i, last_match[:m].to_i, last_match[:d].to_i)
    logger.info("no matching link found")
    exit 0
  end

  logger.info("downloading new file #{name}")
  sleep 1
  blz = agent.get(target_link.href)

  name_match = name.match(/BLZ_(?<y>\d{4})(?<m>\d\d)(?<d>\d\d)\.zip$/)
  target_name = format("../data/%4d_%02d_%02d.tsv.gz", name_match[:y].to_i, name_match[:m].to_i, name_match[:d].to_i)
  target_file = File.expand_path(target_name, __dir__)

  logger.info("extracting ZIP file")
  Zip::File.open_buffer(blz.body).each do |entry|
    next unless entry.name == "BLZ.txt"
    logger.info("found BLZ data")

    entry.get_input_stream do |stream|
      logger.info("reformatting, saving as #{target_file}")
      Zlib::GzipWriter.open(target_file, Zlib::BEST_COMPRESSION) do |gz|
        while line = stream.gets
          line = line.encode(Encoding::UTF_8, Encoding::ISO_8859_15).chomp!
          if line.length != METRIC_LEN
            logger.error("expected line length #{METRIC_LEN}, got #{line.length} in '#{line}'")
            next
          end

          i = 0
          gz.puts METRIC.inject([]) {|s, m| s << line[i ... (i+=m)].strip }.join("\t")
        end
      end
    end
  end

  logger.info("suggestion new version: 0.2.0.#{name_match[:y]}#{name_match[:m]}#{name_match[:d]}")
ensure
  logger.info("logout")
  sleep 1
  agent.get NAVIGATION[:logout]
end
